### PR TITLE
Fixing a race condition in EnrichCoordinatorProxyAction that can leave an item stuck in its queue

### LINKED ISSUE #90598: [CI] CoordinatorTests testAllSearchesExecuted failing
**Build scan:**
https://gradle-enterprise.elastic.co/s/g645ag6b2sko6/tests/:x-pack:plugin:enrich:test/org.elasticsearch.xpack.enrich.action.CoordinatorTests/testAllSearchesExecuted

**Reproduction line:**
`./gradlew ':x-pack:plugin:enrich:test' --tests "org.elasticsearch.xpack.enrich.action.CoordinatorTests.testAllSearchesExecuted" -Dtests.seed=341859B892B7E135 -Dtests.locale=ms-MY -Dtests.timezone=America/Toronto -Druntime.java=17`

**Applicable branches:**
main

**Reproduces locally?:**
Didn't try

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.xpack.enrich.action.CoordinatorTests&tests.test=testAllSearchesExecuted

**Failure excerpt:**
```
java.lang.AssertionError: (No message provided)

  at __randomizedtesting.SeedInfo.seed([341859B892B7E135:A904552DEE08BD63]:0)
  at org.junit.Assert.fail(Assert.java:86)
  at org.junit.Assert.assertTrue(Assert.java:41)
  at org.junit.Assert.assertTrue(Assert.java:52)
  at org.elasticsearch.xpack.enrich.action.CoordinatorTests.testAllSearchesExecuted(CoordinatorTests.java:381)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(NativeMethodAccessorImpl.java:-2)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
  at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  at java.lang.reflect.Method.invoke(Method.java:568)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:44)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:833)

```

### PR DESCRIPTION
**Pull request by masseyke**
There is a race condition in EnrichCoordinatorProxyAction that can result in an item being stuck in its queue even once all threads related to any `schedule()` calls have completed. The item will be flushed out on the next call to `schedule()` but there is no guarantee if or when that will happen. This PR adds an additional check for orphaned items in the queue.

Here's what I believe is happening (I can only reproduce it in fewer than 1 in 10,000 tries so I don't have direct evidence):

- Say thread # 1 calls schedule(), and then coordinateLookups(). It drains the whole queue, and comes through that while loop a second time and there's nothing there so it's about to call remoteRequestPermits.release()
- But while that happens thread # 2 is in schedule() and calls offer.
- So now thread # 1 still has the remoteRequestPermits lock and has decided there's nothing in the queue.
- So now thread # 2 comes into coordinateLookups() and gets false from remoteRequestPermits.tryAcquire() because thread # 1 still has the lock
- Now thread # 2 exists from coordinateLookups() and thread # 1 exits from coordinateLookups() but there's still 1 thing in the queue
- So now there's something in the queue that's just going to stick around until someone schedules something else.

(Note that there are actually more threads than just the 2 I mention since coordinateLookups() makes an async call back to itself)

Closes #90598

### CODE REVIEW

**Comment by masseyke** [comment]
This PR causes a few more loops in the code, but I don't think it will be a noticeable performance hit -- the additional loops are rare and fast. I ran the test (`CoordinatorTests.testAllSearchesExecuted()`) 100,000 times, and this branch was hit fewer than 7,000 times. The runtime was no different than it was without the change.

**Comment by masseyke** [comment]
@elasticmachine update branch

File: x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/action/EnrichCoordinatorProxyAction.java [hunk 1]
```diff
@@ -160,7 +160,17 @@ void coordinateLookups() {
                 final List<Slot> slots = new ArrayList<>(Math.min(queue.size(), maxLookupsPerRequest));
                 if (queue.drainTo(slots, maxLookupsPerRequest) == 0) {
                     remoteRequestPermits.release();
-                    return;
+                    /*
+                     * It is possible that something was added to the queue after the drain and before the permit was released, meaning
+                     * that the other thread could not acquire the permit, leaving an item orphanied in the queue. So we check the queue
```
Comment (jbaiera) [inline x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/action/EnrichCoordinatorProxyAction.java:165]:
```suggestion
                     * that the other thread could not acquire the permit, leaving an item orphaned in the queue. So we check the queue
```

**Review by jbaiera — APPROVED** [review]
LGTM, fixed a small typo is all.

**Review by martijnvg — COMMENTED** [review]
LGTM

**Pull request merged**